csrc/optimization/alias_analysis.cpp

```diff
@@ -103,27 +42,151 @@ class AliasFinder : public OptOutConstDispatch {
+  //  b  |
+  //  e  | C  C     e
+  if (outer_id->hasExpandedExtent() && inner_id->hasExpandedExtent()) {
+    return {true, std::nullopt};
+  }
+  if (outer_id->hasExpandedExtent() || inner_id->hasExpandedExtent()) {
+    return {false, std::nullopt};
+  }
+
+  // o\i | t  f  b  e
+  // ----+-----------
+  //  t  | t  f
+  //  f  | C  C
+  //  b  |
+  //  e  |
+  if (*outer_contiguity) {
+    return {true, inner_contiguity};
+  }
+  return {false, std::nullopt};
+}
+
 void AliasFinder::handle(const ViewOp* view) {
   TensorView* in = view->in();
   TensorView* out = view->out();
 
+  const std::vector<IterDomain*>& in_rfactor = in->getMaybeRFactorDomain();
+  const std::vector<IterDomain*>& out_root = out->getRootDomain();
+  const std::vector<IterDomain*>& out_rfactor = out->getMaybeRFactorDomain();
+
   Layout in_layout = analysis_.preferredLayout(in);
-  const std::vector<IterDomain*>& out_allocation =
-      out->getMaybeAllocationDomain();
-  if (in_layout.allocation_domain == in->getMaybeRFactorDomain() &&
-      isContiguous(in_layout.contiguity) &&
-      out_allocation == out->getMaybeRFactorDomain() && isContiguous(*out) &&
-      !transformsExpandedBroadcastIterDomain(in, out)) {
-    // This is a sufficient but not necessary condition for `out` to alias
-    // `in`. Both `in` and `out` are allocated contiguously per the
-    // rfactor domain. Also, the ViewOp can't transform any expanded broadcast
-    // IterDomain.
-    analysis_.add(
-        out,
-        in,
-        {out_allocation,
-         TensorDomain::getContiguityFilledWith(out_allocation, true)});
+  if (!ir_utils::computePermutation(in_rfactor, in_layout.allocation_domain)
+           .has_value()) {
+    // Give up when `in`'s allocation domain is not an rfactor permutation.
+    return;
   }
+
+  std::unordered_map<IterDomain*, IterDomain*> in_rfactor_to_out_root =
+      PairwiseRootDomainMap(in, out).mapBroadcast(true).mapProducerToConsumer();
+
+  // Collect the allocation order of `in`'s rfactor domain and thus `out`'s root
+  // domain.
+  LinkedHashMap<IterDomain*, std::optional<bool>> allocation_to_contiguity;
+  for (const auto i : c10::irange(in_layout.allocation_domain.size())) {
+    IterDomain* in_allocation_id = in_layout.allocation_domain[i];
+    if (!in_rfactor_to_out_root.count(in_allocation_id)) {
+      // `in_allocation_id` is a reduction product.
+      continue;
+    }
+    IterDomain* out_root_id = in_rfactor_to_out_root.at(in_allocation_id);
+    allocation_to_contiguity.pushBack(out_root_id, in_layout.contiguity[i]);
+  }
+
+  // Replay `Expr`s from `out`'s root to `out`'s rfactor on `out`'s root.
+  // Stop when an `Expr` requires a data copy; otherwise generate the allocation
+  // order of `out`'s rfactor domain and the corresponding contiguity flags.
+  for (Expr* transform : DependencyCheck::getAllExprsBetween(
+           {out_root.begin(), out_root.end()},
+           {out_rfactor.begin(), out_rfactor.end()})) {
```

Comment:
You can use `StmtSort::getExprsBetween` for this.

Reply:
What's the benefit of that? I also saw `Dependencies::getAllExprs`. I'm confused when to use which...

Reply:
@jacobhinkle I merged my PR but would still appreciate your thought on this. I found at least three functions doing the similar work :) I'm not sure when to use which.

Reply:
Great question. It seems to me like `DependencyCheck` and `StmtSort` are redundant; they are used roughly equally often and appear to provide the same functionality, with `DependencyCheck` maybe slightly more complicated. `DependencyCheck` has been around since the beginning of the project, while `StmtSort` evolved from `ExprSort` (https://github.com/csarofeen/pytorch/pull/1376). `ExprSort` was introduced in https://github.com/pytorch/pytorch/pull/45218/files#diff-49787cacb3e793473c24bc615113fb2c782086cbe552c63ac75ffede45f417a1R240-R254. I have been mostly using `StmtSort` which has options for controlling whether we traverse members and attributes. We should make clear in the comments when to use each. CC @naoyam @jjsjann123

Reply:
IIUC, `Dependencies` has been used from a very beginning of the project, and later `IterVisitor::traverseBetween` and `StmtSort` were added. I believe `StmtSort` basically reproduces the same functionality as `Dependencies` but more concisely using `IterVisitor::traverseBetween`. I think it should be valid to replace all `DependencyCheck::getAllExprs/ValsBetween` with `StmtSort::getExprs/Vals`, but should be confirmed by @csarofeen

Reply:
Yes, that's correct.